### DISCUSSION TITLE
The pkg format and array order

### DISCUSSION THREAD

**Discussion by chengm349** [nlohmann/json#4471] [General]
Is it a header only impl? If not, seems I missed the build steps somewhere.

Assume I have 
 "actions": [
        {
          "response": "Accept",
          "delay": 5
        },
        {
          "response": "Accept",
          "delay": null
        }
      ],

Will the parser keep the data order? ie. Delay 5 the first object possibly in an array, Delay null is the 2nd conceptually. This will be my first time to learn JSON and the tool.

I'll use the tool on linux platform with GCC. I quickly went through the supported Compiler and seems no big issue?

**Comment by gregmarr** [comment]
> Is it a header only impl?

https://github.com/nlohmann/json?tab=readme-ov-file#design-goals
>Trivial integration. Our whole code consists of a single header file [json.hpp](https://github.com/nlohmann/json/blob/develop/single_include/nlohmann/json.hpp). That's it. No library, no subproject, no dependencies, no complex build system.

>Will the parser keep the data order? 

Yes and no.  The two actions will stay in the same order as those are array elements, and array elements in JSON are ordered.  The `response` and `delay` will not as those are object elements, and elements of an object in JSON are not ordered.  Since they are not ordered, they are stored by default in a `std::map`, which orders objects by their key values.

This will be stored and dumped like this:
```
"actions": [
  {
    "delay": 5,
    "response": "Accept"
  },
  {
    "delay": null,
    "response": "Accept"
  }
],
```

If you need them to stay with `response` before `delay`, use `nlohmann::ordered_json` instead of `nlohmann::json`.

https://www.json.org/json-en.html
>JSON is built on two structures:
>
>A collection of name/value pairs. In various languages, this is realized as an object, record, struct, dictionary, hash table, keyed list, or associative array.
>An ordered list of values. In most languages, this is realized as an array, vector, list, or sequence.

**Reply by chengm349** [reply]
Yes part is good enough for me. I just need actions in order. inside elements does not matter.

**Comment by chengm349** [comment]
Keen to try.
